packages/api/src/index.ts

```diff
@@ -1,35 +1,68 @@
 export * from './api.gen'
 
-import { ArcadeumAPI as BaseArcadeumAPI } from './api.gen'
 import fetch from 'cross-fetch'
 
-export class ArcadeumAPIClient extends BaseArcadeumAPI {
-  jwtAuth: string | undefined = undefined
+import { ArcadeumAPI as BaseArcadeumAPI } from './api.gen'
+
+export type RefreshJWTCallback = () => Promise<string>
 
-  constructor(hostname: string) {
+export class ArcadeumAPIClient extends BaseArcadeumAPI {
+  constructor(
+    hostname: string,
+    private jwtAuth?: string,
+    private refreshJWT?: RefreshJWTCallback
+  ) {
     super(hostname, fetch)
     this.fetch = this._fetch
   }
 
-  _fetch = (input: RequestInfo, init?: RequestInit): Promise<Response> => {
-    return new Promise<Response>((resolve, reject) => {
-      // automatically include jwt auth header to requests
-      // if its been set on the api client
-      const headers: { [key: string]: any } = {}
-      if (this.jwtAuth && this.jwtAuth.length > 0) {
-        headers['Authorization'] = `BEARER ${this.jwtAuth}`
+  _fetch = async (input: RequestInfo, init?: RequestInit, retry: boolean = true): Promise<Response> => {
+    // automatically include jwt auth header to requests
+    // if its been set on the api client
+    const headers: { [key: string]: any } = {}
+    if (this.jwtAuth && this.jwtAuth.length > 0) {
+      headers['Authorization'] = `BEARER ${this.jwtAuth}`
+    }
+
+    // before the request is made
+    init!.headers = { ...init!.headers, ...headers }
+
+    // we should avoid retrying if the request is to /GetAuthToken itself
+    let url: string
+    switch (typeof input) {
+      case 'string':
+        url = input
+        break
+      case 'object':
+        url = input.url
+        break
+    }
+    const isGetAuthToken = new URL(url).pathname.endsWith('/GetAuthToken')
```

Comment:
I'm not sure if this `_fetch` could ever get called with user-supplied URLs, but if it could, there's no guarantee that the URL passed to a fetch is a "full" URL, it might just be a path "/test" or something else.

One solution is to wrap in try/catch.

Reply:
In the generated bindings, it always uses fully resolved URLs: https://github.com/0xsequence/sequence.js/blob/1a3c9f43b38b94b87af1a47bd537ea9ff92f7479/packages/api/src/api.gen.ts#L812, but I'll make this more robust so it doesn't have to assume that.